[ARM] Add instruction selection for strict FP

Varnike opened the pull request:
This consists of marking the various strict opcodes as legal, and adjusting instruction selection patterns so that 'op' is 'any_op'. The changes are similar to those in D114946 for AArch64.

Custom lowering and promotion are set for some FP16 strict ops to work correctly.

This PR is part of the work on adding strict FP support in ARM, which was previously discussed in #137101.

Varnike committed:
[ARM] Add instruction selection for strict FP

This consists of marking the various strict opcodes as legal, and
adjusting instruction selection patterns so that 'op' is 'any_op'.
The changes are similar to those in D114946 for AArch64.

Custom lowering and promotion are set for some FP16 strict ops to
work correctly.

davemgreen reviewed (commented):
Can you rebase?

I was going to say we could do with extra testing. It looks like there were some added in llvm/test/CodeGen/ARM/fp-intrinsics.ll, can any part be un-fixme'd now?

davemgreen reviewed (commented):
Thanks this looks decent to me, I just have some comments on the tests.

davemgreen reviewed (commented):
Thanks could we change the other `-mtriple arm--` too?

Otherwise LGTM, are you happy for this to be submitted?

Varnike committed:
Change tests triple from arm--- to armv7--

Varnike commented:
Sorry for the delay. Corrected test triples and added lowering for STRICT_FP16_TO_FP in cases where no hardware implementation is available.
> Otherwise LGTM, are you happy for this to be submitted?

If these changes are correct for you, then yes.

davemgreen reviewed (approved):
OK Great thanks. LGTM

If looks like there is a conflict in llvm/lib/Target/ARM/ARMISelLowering.cpp that needs to be addressed before we can submit.

Varnike committed:
Post-rebase update

Varnike committed:
Add several strict-fp tests

Varnike committed:
Post-rebase update

Varnike committed:
Correct strict-fp-int-promote test
